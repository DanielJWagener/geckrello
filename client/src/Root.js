// Simply creates a Redux store and passes it in to a provider tag.
// This logic was originally in index.js, but we've movied it here to
// make our testing process easier.
import React from "react";
import { Provider } from "react-redux";
import { createStore, applyMiddleware, compose } from "redux";
import reduxThunk from "redux-thunk";
import reducers from "./reducers";

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const store = createStore(
  reducers,
  {},
  composeEnhancers(applyMiddleware(reduxThunk))
);

export default ({ children }) => {
  return <Provider store={store}>{children}</Provider>;
};
