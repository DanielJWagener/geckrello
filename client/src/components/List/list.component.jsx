import AddCard from "../AddCard/add-card.component";
import Card from "../Card/card.component";
import CardModal from "../CardModal/card-modal.component";
import React from "react";
import { connect } from "react-redux";
import { archiveList } from "../../actions";
import ListDropTarget from "./list-drop-target";

import "./list.styles.scss";

export const List = props => {
  // Event handler for when user clicks the remove button
  const archiveList = () => {
    props.archiveList(props.listId);
  };

  // Make array of Card components from props
  const cardsArray = () =>
    props.cards.map(card => (
      <Card
        key={card._id}
        cardId={card._id}
        cardTitle={card.title}
        inList={props.listTitle}
        description={card.description}
      />
    ));

  //Make array of CardPopup components from props
  const cardModalsArray = () =>
    props.cards.map(card => (
      <CardModal
        key={card._id}
        cardId={card._id}
        cardTitle={card.title}
        inList={props.listTitle}
      />
    ));

  return (
    <ListDropTarget listId={props.listId}>
      <div className="list">
        <div className="list__remove" onClick={archiveList}>
          &times;
        </div>
        <h2 className="list__heading">{props.listTitle}</h2>
        <div className="list__cards">{cardsArray()}</div>
        <AddCard listHome={props.listId} />
      </div>
      <div className="card-modals">{cardModalsArray()}</div>
    </ListDropTarget>
  );
};

const mapStateToProps = (state, ownProps) => {
  //Iterate over every card in state, return only the non-archived ones with a listHome corresponding to this list
  const { listId } = ownProps;
  const cards = state.cards.filter(
    card => card.listHome === listId && !card.archived
  );

  return { cards };
};
export default connect(mapStateToProps, { archiveList })(List);
