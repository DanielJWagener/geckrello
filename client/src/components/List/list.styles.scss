@import "../../sass/abstracts/_variables.scss";

.list {
  width: 100%;
  background-color: $component-background;
  border-radius: 0.5rem;
  min-height: 120px;
  position: relative;
  margin-bottom: 2rem;
  z-index: 0;

  &__heading {
    font-size: 2rem;
    width: 90%;
    overflow: hidden;
    font-family: sans-serif;
    padding: 1rem 1rem 1.5rem 1rem;
  }

  &__remove {
    position: absolute;
    top: 3px;
    right: 1rem;
    cursor: pointer;
    font-size: 3rem;

    &:hover {
      color: red;
    }
  }

  &__cards {
    padding: 0.5rem 0.5rem 0.5rem 0.5rem;
  }

  &:after {
    content: "";
    position: absolute;
    top: 0;
    left: 0;
    bottom: 0;
    right: 0;
    z-index: -1;
    box-shadow: 0 1.5rem 4rem rgba(0, 0, 0, 0.3);
  }
}
