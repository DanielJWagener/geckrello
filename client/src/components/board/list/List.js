import AddCard from "./AddCard";
import Card from "./card/Card";
import CardModal from "./card/CardModal";
import React from "react";
import { connect } from "react-redux";
import { archiveList } from "../../../actions";
import ListDropTarget from "./ListDropTarget";
import memoize from "memoize-one";

class List extends React.Component {
  state = {
    spans: 13,
    height: 130,
    addCardHeight: 67
  };

  // Element height handlers. The Lists' height is based on CSS grid spans, so when the Lists' height changes, we have to set a new value for the spans.
  headingRef = React.createRef();
  cardsRef = React.createRef();

  setSpansListToggle = addCardHeight => {
    // The height of the List is the sum of the heights of its heading, its cards div and the AddCard component. Since we have to create a ref for AddCard directly on that component, we make this function a prop of AddCard, retrieve the height from inside that AddCard, and pass that height as an argument to this function.
    const height =
      this.headingRef.current.clientHeight +
      this.cardsRef.current.clientHeight +
      addCardHeight;

    const spans = Math.ceil(height / 10) + 1; // +1 to ensure a minimum bottom margin of one span height for the List

    this.setState({ spans: spans, height: height });
    if (addCardHeight !== this.state.addCardHeight) {
      this.setState({ addCardHeight: addCardHeight });
    }
  };

  setSpansAddCard = addCardHeight => {
    // When adding a card, the cards div's ref gives us its height before the card has been added. Therefore, before the card is added, we need to predict what that div's height will be and set the List's height accordingly.
    let cardsHeight;
    const cardsDivHeightWhenEmpty = 10;
    const heightOfOneCard = 46;

    if (this.cardsRef.current.clientHeight === cardsDivHeightWhenEmpty) {
      cardsHeight = heightOfOneCard;
    } else {
      cardsHeight = this.cardsRef.current.clientHeight + heightOfOneCard;
    }

    const height =
      this.headingRef.current.clientHeight + cardsHeight + addCardHeight;

    const spans = Math.ceil(height / 10) + 1;
    this.setState({ spans: spans, height: height });
    if (addCardHeight !== this.state.addCardHeight) {
      this.setState({ addCardHeight: addCardHeight });
    }
  };

  // Function to handle list height change, memoized to respond to drag-and-drop events (i.e. to cards prop changes)
  setSpansDnd = memoize((cardsArray, addCardHeight) => {
    const heightOfOneCard = 46;
    const headingHeight = 48;
    const height =
      headingHeight + cardsArray.length * heightOfOneCard + 5 + addCardHeight;
    const spans = Math.ceil(height / 10) + 1;

    this.setState({ spans: spans, height: height });
  });

  // Event handler for when user clicks the remove button
  archiveList = () => {
    this.props.archiveList(this.props.listId);
  };

  // Make array of Card components from props
  cardsArray = () =>
    this.props.cards.map(card => (
      <Card
        key={card.cardId}
        cardId={card.cardId}
        cardTitle={card.title}
        inList={this.props.listTitle}
      />
    ));

  // Make array of CardPopup components from props
  cardModalsArray = () =>
    this.props.cards.map(card => (
      <CardModal
        key={card.cardId}
        cardId={card.cardId}
        cardTitle={card.title}
        inList={this.props.listTitle}
      />
    ));

  render() {
    // Memoized function: only gets called when value of one of the arguments changes (essentially, when a card gets added to or removed from props)
    this.setSpansDnd(this.props.cards, this.state.addCardHeight);

    return (
      <ListDropTarget
        // Style ensures drop target is the same size as the list
        style={{
          height: `${this.state.height}px`,
          width: "100%",
          gridRowEnd: `span ${this.state.spans}`
        }}
        listId={this.props.listId}
      >
        <div
          className="list"
          ref={this.listRef}
          style={{ gridRowEnd: `span ${this.state.spans}` }}
        >
          <div className="list__remove" onClick={this.archiveList}>
            &times;
          </div>
          <h2 className="list__heading" ref={this.headingRef}>
            {this.props.listTitle}
          </h2>
          <div className="list__cards" ref={this.cardsRef}>
            {this.cardsArray()}
          </div>
          <AddCard
            listHome={this.props.listId}
            setSpansUpdate={this.setSpansListToggle}
            setSpansUpdateForCard={this.setSpansAddCard}
          />
        </div>
        <div className="card-modals">{this.cardModalsArray()}</div>
      </ListDropTarget>
    );
  }
}

const mapStateToProps = (state, ownProps) => {
  //Iterate over every card in state, return only the non-archived ones with a listHome corresponding to this list
  const { listId } = ownProps;
  const cards = state.cards.filter(
    card => card.listHome === listId && !card.archived
  );

  return { lists: state.lists, cards: cards };
};
export default connect(
  mapStateToProps,
  { archiveList }
)(List);
