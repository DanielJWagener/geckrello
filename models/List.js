const mongoose = require("mongoose");
const { Schema } = mongoose;

const listSchema = new Schema({
  title: String,
  boardHome: { type: mongoose.Schema.ObjectId, ref: "boards" }
});

const List = mongoose.model("lists", listSchema);

module.exports = List;
