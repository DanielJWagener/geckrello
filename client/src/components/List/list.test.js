import { shallow } from "enzyme";
import React from "react";
import List from "./list.component";
import Root from "../../Root";

it("renders List component", () => {
  expect(
    shallow(
      <Root>
        <List />
      </Root>
    )
  ).toMatchSnapshot();
});
