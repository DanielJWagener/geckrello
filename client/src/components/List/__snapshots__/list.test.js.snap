// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders List component 1`] = `ShallowWrapper {}`;
